New Resource: aws_rds_instance

HackerShark opened the pull request:
As per issue #164 the following is the skeleton for the RDS Instance Resource.

- Adds in functionality to check if an RDS instance should or should not exist

HackerShark committed:
Adding RDS API to connections

Signed-off-by: HackerShark <melsharkawi@mitre.org>

HackerShark committed:
Adding skeleton resource for AWS RDS Instance

Signed-off-by: HackerShark <melsharkawi@mitre.org>

HackerShark committed:
Created terraform to generate rds instance in aws

Signed-off-by: HackerShark <melsharkawi@mitre.org>

HackerShark committed:
Adding skeleton tests for an RDS instance if it should and should not exist

Signed-off-by: HackerShark <melsharkawi@mitre.org>

HackerShark committed:
Adding unit tests for the AWS RDS instance resource

Signed-off-by: HackerShark <melsharkawi@mitre.org>

HackerShark committed:
Adding skeleton docs for the AWS RDS Instance resource

Signed-off-by: HackerShark <melsharkawi@mitre.org>

samcornwell committed:
aws_vpc resource skeleton (#172)

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>
Signed-off-by: Sam Cornwell <14048146+samcornwell@users.noreply.github.com>

HackerShark committed:
Fixed some styling issues

Signed-off-by: HackerShark <melsharkawi@mitre.org>

clintoncwolfe commented on libraries/aws_rds_instance.rb:
```diff
@@ -0,0 +1,75 @@
+  def fetch_from_aws
+    backend = AwsRdsInstance::BackendFactory.create
+
+    # Transform into filter format expected by AWS
```
This block - from line 42 to line 48 - is unused, and can be deleted.  It's copied from other resources in which the AWS SDK expects a `filter` option, for which we have to construct criteria from our instance variables.

While describe_db_instances does take a filter option, it's not being used here, and we should remove code that is unused.  We can easily add it back in if we expand the interface.

clintoncwolfe commented on test/integration/default/build/rds.tf:
```diff
@@ -0,0 +1,80 @@
+  skip_final_snapshot  = true
+}
+
+output "rds_db_instance_address" {
```
Why are you outputting so many things?  You're only using `rds_db_instance_id` in the integration test.

clintoncwolfe commented on test/unit/resources/aws_rds_instance_test.rb:
```diff
@@ -0,0 +1,83 @@
+    AwsRdsInstance.new('test-instance-id')
+  end
+
+  def test_constructor_expected_well_formed_args
```
This test should only be used when there are combinations of parameters that are valid (and other combinations that are not).  Here, we have a simple scalar ID pattern; I don't think this is needed.

clintoncwolfe commented on test/unit/resources/aws_rds_instance_test.rb:
```diff
@@ -0,0 +1,83 @@
+  end
+end
+
+#=============================================================================#
```
A section demonstrating search / recall is needed.

clintoncwolfe commented on test/unit/resources/aws_rds_instance_test.rb:
```diff
@@ -0,0 +1,83 @@
+    end
+  end
+
+  class Basic < AwsRdsInstance::Backend
```
This is currently unused, but is needed for a search / recall section.

clintoncwolfe commented on test/unit/resources/aws_rds_instance_test.rb:
```diff
@@ -0,0 +1,83 @@
+    def describe_db_instances(query)
+      fixtures = [
+          OpenStruct.new({
+                             db_id: 'some-db',
```
While you don't have to include all properties returned by the AWS API, you do have to match the property names exactly for the test to work.  Here, you want `db_instance_identifier`, not `db_id`.

clintoncwolfe commented on test/unit/resources/aws_rds_instance_test.rb:
```diff
@@ -0,0 +1,83 @@
+                         }),
+      ]
+
+      selected = fixtures.select do |db|
```
This won't work.  Right now, the resource is querying using the key `db_instance_identifier`, and not sending a `filter` at all.  This code block only uses filter, and so will never find a record.

clintoncwolfe commented on libraries/aws_rds_instance.rb:
```diff
@@ -0,0 +1,75 @@
+  "
+
+  include AwsResourceMixin
+  attr_reader :db_id
```
This property should be named in line with the AWS API - `db_instance_identifier`.

samcornwell committed:
skeletal aws_vpcs (#182)

Signed-off-by: Sam Cornwell <14048146+samcornwell@users.noreply.github.com>

clintoncwolfe committed:
aws_s3_bucket with modified interface (#183)

    Signed-off-by: Matthew Dromazos <dromazmj@dukes.jmu.edu>
    Signed-off-by: Aaron Lippold <lippold@gmail.com>
    Signed-off-by: Sam Cornwell <14048146+samcornwell@users.noreply.github.com>
    Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

aaronlippold committed:
README clarifications about creating a profile that relies on the inspec-aws resource pack (#153)

Signed-off-by: Aaron Lippold <lippold@gmail.com>

samcornwell committed:
added require statement for _aws.rb for all resources (#199)

Signed-off-by: Sam Cornwell <14048146+samcornwell@users.noreply.github.com>

clintoncwolfe committed:
Remove highline and nokogiri, and group test gems in Gemfile (#203)

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

rx294 committed:
Correct access_key detection on aws_iam_root_user  (#198)

Signed-off-by: Rony Xavier <rx294@nyu.edu>

rx294 committed:
Skeletal aws_cloudtrail_trails resource (#191)

Signed-off-by: Rony Xavier <rx294@nyu.edu>

rx294 committed:
Functional aws_cloudtrail_trail resource (#186)

Signed-off-by: Rony Xavier <rx294@nyu.edu>

clintoncwolfe committed:
Make cloudwatch log metric filter test fixtures more unique (#201)

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

rx294 committed:
aws_iam_policy resource (#184)

Signed-off-by: Rony Xavier <rx294@nyu.edu>
Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

miah committed:
Ensure unique security group name while integration testing. (#218)

Signed-off-by: Miah Johnson <miah@chia-pet.org>

clintoncwolfe committed:
Remove accidentally committed notes file (#194)

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Silence some test warnings (#140)

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

rx294 committed:
aws_iam_access_keys incorrectly populates created_date (#215)

Signed-off-by: Rony Xavier <rx294@nyu.edu>

rx294 committed:
skeletal aws_iam_ policies resource (#193)

Signed-off-by: Rony Xavier <rx294@nyu.edu>

clintoncwolfe committed:
Skeletal aws_kms_keys resource

Signed-off-by: Rony Xavier <rx294@nyu.edu>
Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

rx294 committed:
Password usage properties for aws_iam_users (#213)

Signed-off-by: Rony Xavier <rx294@nyu.edu>

rx294 committed:
Add have_access_logging_enabled matcher to aws_s3_bucket (#212)

Signed-off-by: Rony Xavier <rx294@nyu.edu>

samcornwell committed:
Skeletal aws_iam_groups resource (#208)

Signed-off-by: Sam Cornwell <14048146+samcornwell@users.noreply.github.com>

samcornwell committed:
Skeletal aws_iam_group resource (#221)

Signed-off-by: Sam Cornwell <14048146+samcornwell@users.noreply.github.com>

aaronlippold committed:
Add 'require _aws' to newly added resources (#224)

Signed-off-by: Aaron Lippold <lippold@gmail.com>

dromazmj committed:
Skeletal Resource: aws_vpc_subnet (#209)

Signed-off-by: Matthew Dromazos <dromazmj@dukes.jmu.edu>

miah committed:
Skeletal Resource: aws_route_table (#217)

Signed-off-by: Miah Johnson <miah@chia-pet.org>

HackerShark committed:
Updating code after PR review.

aws_rds_instance.md:
  - Changed wording from `RDS` to `RDS instance`

libraries/aws_rds_instance.rb:
  - changed `db_id` to `db_instance_identifier`
  - removed unused code
  - changed from catchall of StandardError to catching the specific DBInstanceNotFound error

rds.tf:
  - removed unused outputs

aws_rds_instance_test.rb:
  - changed `db_id` to `db_instance_identifier`
  - Added Search/Recall section
  - removed unused code

Signed-off-by: HackerShark <melsharkawi@mitre.org>

HackerShark committed:
Merge branch 'master' into aws_rds_instance

clintoncwolfe commented:
As part of the move to the inspec core repo, this PR is being closed.  We've had some private messages about extra commits and merges on this branch.  I really want to get your valuable contribution, so I suggest that you get a fresh clone of the `inspec` repo, then copy in your files from this branch and add them individually (it should be about 5 - a TF file and a test control for integration testing, a docs file, a unit test, and the resource code itself).  You'll need to look around and note where things have moved:
- Resource files moved from `libraries` to `lib/resources/aws`
- You'll need to add a require for your resource to `lib/resource_support/aws.rb`
- Integration tests moved from `test/integration` to `test/aws`
- Several changes in the way the resources connect to AWS (now relying on train, instead of AwsConnection).  I can help you with that after you open a PR on inspec. 

Thanks again for all your work on this!
